### ISSUE TITLE
HashSyntax: autocorrect Under Ruby 2.2 produces code that no longer works on Ruby 2.0

### ISSUE THREAD

**Issue by dblock** [bbatsov/rubocop#2389]
Is this by design? Many projects build on all kinds of rubies, so this behavior is undesirable.

On https://github.com/intridea/hashie

```
git checkout acc3d48b91bb318d91be63f116a1f252dbfae636

$ ((acc3d48...))$ git status
HEAD detached at acc3d48
nothing to commit, working directory clean

$ ((acc3d48...))$ bundle
Using rubocop 0.34.2 (was 0.27.1)
...

$ ((acc3d48...))$ rubocop -a
Inspecting 59 files
...........................................................

59 files inspected, no offenses detected

$ ((acc3d48...))$ rvm list
...
=* ruby-2.0.0-p598 [ x86_64 ]

$ ((acc3d48...))$ rvm use 2.2
Using /Users/dblock/.rvm/gems/ruby-2.2.0

$ ((acc3d48...))$ bundle
...

$ ((acc3d48...))$ rubocop -a
warning: parser/current is loading parser/ruby22, which recognizes
warning: 2.2.3-compliant syntax, but you are running 2.2.0.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 59 files
...................................................C.......

Offenses:

spec/hashie/hash_spec.rb:44:35: C: [Corrected] Use the new Ruby 1.9 hash syntax.
    expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3])
                                  ^^^^^
spec/hashie/hash_spec.rb:44:48: C: [Corrected] Use the new Ruby 1.9 hash syntax.
    expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3])
                                               ^^^^^^^^^
spec/hashie/hash_spec.rb:44:65: C: [Corrected] Use the new Ruby 1.9 hash syntax.
    expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3])
                                                                ^^^^^^^^^
spec/hashie/hash_spec.rb:81:37: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                    ^^^^^
spec/hashie/hash_spec.rb:81:50: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                 ^^^^^^^^^
spec/hashie/hash_spec.rb:81:67: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                                  ^^^^^^^^^
spec/hashie/hash_spec.rb:81:99: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                                                                  ^^^^^
spec/hashie/hash_spec.rb:81:112: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                                                                               ^^^^^
spec/hashie/hash_spec.rb:81:125: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                                                                                            ^^^^^^^^^
spec/hashie/hash_spec.rb:81:142: C: [Corrected] Use the new Ruby 1.9 hash syntax.
      expect(symbolized_hash).to eq(:a => 'hey', :"123" => 'bob', :array => [1, 2, 3], subhash: { :a => 'hey', :b => 'bar', :'123' => 'bob', :array => [1, 2, 3] })
                                                                                                                                             ^^^^^^^^^

59 files inspected, 10 offenses detected, 10 offenses corrected
```

This produces code that no longer works on 2.0.

```
$ ((acc3d48...))$ rvm use 2.0
Using .../ruby-2.0.0-p598
rspec spec
$ ((acc3d48...))$ rspec spec
.../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load': /Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:44: syntax error, unexpected ':', expecting => (SyntaxError)
...ed_hash).to eq(a: 'hey', "123": 'bob', array: [1, 2, 3])
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:44: syntax error, unexpected ',', expecting keyword_end
...).to eq(a: 'hey', "123": 'bob', array: [1, 2, 3])
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:44: syntax error, unexpected ')', expecting keyword_end
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected ':', expecting =>
...ed_hash).to eq(a: 'hey', "123": 'bob', array: [1, 2, 3], sub...
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected ',', expecting keyword_end
...).to eq(a: 'hey', "123": 'bob', array: [1, 2, 3], subhash: {...
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected ',', expecting keyword_end
..."123": 'bob', array: [1, 2, 3], subhash: { a: 'hey', b: 'bar...
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected ':', expecting =>
...h: { a: 'hey', b: 'bar', '123': 'bob', array: [1, 2, 3] })
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected ',', expecting keyword_end
... 'hey', b: 'bar', '123': 'bob', array: [1, 2, 3] })
...                               ^
/Users/dblock/source/hashie/dblock/spec/hashie/hash_spec.rb:81: syntax error, unexpected '}', expecting keyword_end
...123': 'bob', array: [1, 2, 3] })
...                               ^
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `block in load_spec_files'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `each'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/configuration.rb:1105:in `load_spec_files'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:96:in `setup'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:84:in `run'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:69:in `run'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/lib/rspec/core/runner.rb:37:in `invoke'
  from .../ruby-2.0.0-p598/gems/rspec-core-3.1.7/exe/rspec:4:in `<top (required)>'
  from .../ruby-2.0.0-p598/bin/rspec:23:in `load'
  from .../ruby-2.0.0-p598/bin/rspec:23:in `<main>'
  from .../ruby-2.0.0-p598/bin/ruby_executable_hooks:15:in `eval'
  from .../ruby-2.0.0-p598/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by alexdowad** [comment]
When correcting hash syntax, RuboCop does a check like this to make sure it doesn't break the code:

```ruby
RuboCop::ProcessedSource.new("{ #{sym_name}: :foo }").valid_syntax?
```

In other words, it parses that snippet and makes sure the syntax is valid. **But**, here is the problem: whatever version of the Ruby interpreter you are running RuboCop on, determines the parser which it uses to check that code.

This is obviously problematic for libraries which aim to run on more than one version of Ruby (which is to say, all of them).

I suggest that RuboCop should have a global config setting which represents the Ruby versions (1.9, 2.0, 2.1, and so on) which the code is expected to run on. When parsing sample code, it should try with *all* of the listed versions (using `Parser::Ruby19`, `Parser::Ruby20`, and so on).

How does this sound to the maintainers?

**Comment by bbatsov** [comment]
It sounds reasonable. 

The behaviour in question was implemented as it is because doing it with regular expressions was kind of problematic. I definitely think it makes sense to be able to force the version of the parser used by RuboCop.

**Comment by alexdowad** [comment]
> The behaviour in question was implemented as it is because doing it with regular expressions was kind of problematic.

Absolutely. Since we have a Ruby parser available, nothing is wrong with reusing it to check if autocorrection is valid or not. That is good design.

The problem is just the dependency on the current Ruby version.

After more thought, I don't think it will be necessary to test the corrected code on multiple Ruby versions... just the **oldest** Ruby version which the user wants to support.
